[Adam Oh Zhi Hong] iP

adamozh opened the pull request:
# Duke - Your personal assistant chatbot

> Why do anything unless it's going to be great?

Duke is not just any chatbot - it's **your** chatbot which helps you keep your life in check!
Track tasks like never before with its checklist and categorical system.

## Features
* Tracks Todos, Deadlines, Events
* Mark tasks as done
* Search for tasks
* Delete tasks

## Getting Started
1. Download the latest release [here](https://github.com/moreTriangles/ip)
1. Simply double click the jar file to get started!

## New Updates
Latest updates! 😉
- [x] New updated GUI
- [x] Profile Pictures
- [x] Shorthand Notation

Commit:
Add Gradle support

Commit:
Ignore Java class files

Commit:
Implement skeletal functions: Greet, Echo, Exit

Commit:
Create Task.java

Commit:
Create TaskManager.java

Commit:
Fix bug in TaskManager and Task

Commit:
Fix typos in TaskManager.java

Commit:
Implement Add and List functionality, added some spacing

Commit:
Add isDone class variable to Task

Commit:
Add done field in list

Commit:
Add setter for isDone

Commit:
Add Mark as Done functionality and simplified Duke responses

Commit:
Created Task subclasses and implemented them in Duke

* Task was made abstract, and instances were replaced by new subclasses
* Generic tasks that are not Todo, Deadline and Event are now rejected by Duke
* Task Manager can now handle all 3 types of tasks and extract necessary
  data.

Commit:
Remove unused function

Commit:
Change phrasing and handle missing /by, /at

Commit:
Add Javadocs

Commit:
Update user prompt to be in new line

Commit:
Set up TextUiTesting

Commit:
Create custom exceptions, implement error handling

Commit:
Add MissingTaskNumberException

Commit:
Add Javadoc for MissingTaskNameException

Commit:
Implement delete, print number of tasks after adding

adamozh committed:
Create /data/duke.txt

adamozh committed:
Overload constructors for Task and subclasses

adamozh committed:
Add data/duke.txt to .gitignore

adamozh committed:
Add data folder to .gitignore and stop tracking it

adamozh committed:
Add save to file functionality

adamozh committed:
Add DukeIOManager to handle file interactions

adamozh committed:
Handle errors and exceptions for file interactions

adamozh committed:
Store dates as LocalDate for Deadline and Event

adamozh committed:
Handle DateTimeParseException

adamozh committed:
Rename variables and getters

adamozh committed:
Merge branch 'branch-Level-7'
Merge branch-Level-7 to master for save session function

adamozh committed:
Merge branch-Level-8 to master for DateTime function

adamozh committed:
Rename DukeIOManager to Storage

adamozh committed:
Create Parser class, rename TaskManager to TaskList

Changes to TaskList (previously TaskManager)
* Make all methods and taskArrayList static
* Add methods for remove, get and size
* Add start method to clear taskArrayList when Duke is started
* Handle function is now done by Parser class (also static)

adamozh committed:
Reorganize Duke.java, remove static property for TaskList

adamozh committed:
Create UI class and make Storage non-static

adamozh committed:
Fix switch-case indentations

adamozh committed:
Abstract print statements to UI class, and refactor code

adamozh committed:
Refactor classes into packages

adamozh committed:
Set up JUnit and add tests for Deadline and Todo

adamozh committed:
Add JavaDocs and clean up code

adamozh committed:
Follow Java coding standards

adamozh committed:
Add find by task name function

adamozh committed:
Merge branch 'branch-A-JavaDoc'
Merge branch-A-Javadoc to master

adamozh committed:
Merge branch-A-CodingStandard to master

adamozh committed:
Merge branch 'branch-Level-9'
Merge branch-Level-9 to master

adamozh committed:
Merge branch 'add-gradle-support'

Merge add-gradle-support to master for gradle support

adamozh committed:
Fix error due to wrong naming and move Duke into package

adamozh committed:
Add Gradle run, build and unit tests

adamozh committed:
Merge branch 'branch-A-Gradle'
Merge branch-A-Gradle to master for Gradle run, build and unit test

adamozh committed:
Fix comments and styling using Checkstyle

adamozh committed:
Merge branch 'branch-A-CheckStyle'
Merge branch-A-CheckStyle to master

adamozh committed:
Add basic GUI, Duke echos the user

adamozh committed:
Add background color, padding, circle clipping

adamozh committed:
Use FXML, create Main class

adamozh committed:
Connect GUI to Duke responses

adamozh committed:
Remove isRunning property

adamozh committed:
Remove readCommand method

adamozh committed:
Tidy code

adamozh committed:
Add welcome message

adamozh committed:
Fix bug where long text automatically collapses

adamozh committed:
Merge branch 'branch-Level-10'
Merge branch-Level-10 to master

moreTriangles commented:
# Duke - Your personal assistant chatbot

> Why do anything unless it's going to be great?

Duke is not just any chatbot - it's **your** chatbot which helps you keep your life in check!
Track tasks like never before with its checklist and categorical system.

## Features
* Tracks Todos, Deadlines, Events
* Mark tasks as done
* Search for tasks
* Delete tasks

## Getting Started
1. Download the latest release [here](https://github.com/moreTriangles/ip)
1. Simply double click the jar file to get started!

## New Updates
Updates that you can look forward to! 😉
- [x] New updated GUI
- [x] Profile Pictures
- [ ] Reminders

That's all for now, watch this space for more updates!

adamozh committed:
Handle 'hi' and 'hello'

adamozh committed:
Fix javadoc and update README.md

HangZelin commented on src/main/java/duke/Parser.java:
```diff
@@ -47,30 +44,21 @@ public static void handle(String userString, TaskList taskList) throws DukeExcep
-            UI.bye();
-            isRunning = false;
-            break;
+            return UI.bye();
```
You can actually instantiate a ui object to help you do the work, and you don't have to make every method in UI static.

adamozh replied:
Yes you're right, I will change that.

HangZelin commented on src/main/java/Duke.java:
```diff
@@ -37,7 +37,22 @@ public static void main(String[] args) {
                     break;
                 default:
-                    tm.handle(command);
+                    try {
```
You don't actually have to define so many Exception types. You can try create a method in DukeException to costumize Duke exception message.

adamozh replied:
Is that considered good practice? I was worried that might mean that a DukeException could have multiple meanings, making it ambiguous.

HangZelin commented on src/main/java/Parser.java:
```diff
@@ -5,83 +5,118 @@
-            tasklist.add(new ToDo(remaining));
-            System.out.println("Great! I've added your todo. Enter 'list' to see your tasks!");
-            System.out.printf("You currently have %d tasks.%n", tasklist.size());
+            handleTodo();
```
I like you encasuplating all these codes into a method to make your code simplier!

adamozh replied:
Thanks! 👍

HangZelin commented on src/main/java/duke/UI.java:
```diff
@@ -54,22 +55,26 @@ public static String added(String taskType) {
+     * @return String of tasks
      */
     public static String list(TaskList taskList) {
         String list = "Here are your current tasks:\n";
```
Is it better to make this String a constant value? which is naming it as LIST and static final?

adamozh replied:
I don't think I can do that because after that there is some string concatenation. Good spot though! I actually haven't been using any constant names because I can't find where to use them.

HangZelin reviewed (commented):
Good job overall! I can see you putting a lot of effort trying to make your code more cleaner and applied to coding standard.

Just some minor issues and improvements for you to take care.

Keep on your good work! 👍

adamozh committed:
Make window non-resizable

adamozh committed:
Handle scenario when there is no tasks

adamozh committed:
Fix task not saving bug

adamozh committed:
Add assert statements

CheongYeeMing commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,51 @@
+import duke.exceptions.MissingTaskNameException;
+import duke.exceptions.MissingTaskNumberException;
+
+/**
+ * Duke is a Personal Assistant Chatbot that keeps track of various tasks
+ *
+ * @author Adam Oh Zhi Hong
+ */
```
Perhaps you can add the `@version` into your class comment, to tell you at which stage the class was last updated.

adamozh replied:
That could be useful, will do that next update!

CheongYeeMing commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,51 @@
+     *
+     * @return String response from Duke
+     */
+    public String getResponse(String input) {
+        String dukeResponse;
+        try {
+            dukeResponse = Parser.handle(input, taskList);
+        } catch (InvalidCommandException e) {
+            dukeResponse = "I'm afraid I don't recognise that, please try again!";
+        } catch (MissingTaskNameException e) {
+            dukeResponse = "Task name cannot be empty!";
+        } catch (MissingTaskNumberException e) {
+            dukeResponse = "Did you forget to enter your task number?";
+        } catch (InvalidTaskNumberException e) {
+            dukeResponse = "Sorry, that task does not exist!";
+        } catch (MissingDeadlineException e) {
+            dukeResponse = "When is that due? Let me know after '/by'!";
+        } catch (MissingEventTimeException e) {
+            dukeResponse = "When is the event happening? Let me know after '/at'!";
+        } catch (DateTimeParseException e) {
+            dukeResponse = "Oops, did you enter your date in yyyy-mm-dd format?";
+        } catch (DukeException e) {
+            dukeResponse = UI.dukeException();
+        }
+        return dukeResponse;
+    }
```
A neat way to handle all the different exceptions simultaneously. 👍

adamozh replied:
Thanks! 👍

CheongYeeMing commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,185 @@
+        String firstWord = arr[0].toLowerCase();
+        Parser.words = arr;
+
+        switch (firstWord) {
+        case "hi":
+            return UI.hi();
+        case "hello":
+            return UI.hi();
+        case "bye":
+            taskList.saveTasksToStorage();
+            return UI.bye();
+        case "list":
+            return UI.list(taskList);
+        case "done":
+            return handleDone();
+        case "todo":
+            return handleTodo();
+        case "deadline":
+            return handleDeadline();
+        case "event":
+            return handleEvent();
+        case "delete":
+            return handleDelete();
+        case "find":
+            return handleFind();
+        default:
+            throw new InvalidCommandException("Invalid command");
```
A neat way to handle all the different commands simultaneously. 👍

adamozh replied:
Thanks! 👍

CheongYeeMing commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,185 @@
+        }
+    }
+
+    /**
+     * Marks the given task as done
+     * Triggered when the done command is identified
+     *
+     * @throws DukeException If there is no task number or wrong task number
+     */
+    public static String handleDone() throws DukeException {
```
Comments missing `@return`. This is true for the following methods as well:

- `handleDone`
- `handleTodo`
- `handleDeadline`
- `handleEvent`
- `handleDelete`
- `handleFind`

adamozh replied:
Thanks for spotting that out! Will fix it the next update.

CheongYeeMing commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,185 @@
+ * Makes sense of the user command
+ * Bridges communication between the user string and the tasks
+ */
+public class Parser {
```
I realised that you made use of a lot of static methods. Why not try instantiating `Ui` objects?

adamozh replied:
Hmm yes, will change that too 👍

CheongYeeMing reviewed (commented):
Well written code in general. Not sure if it's necessary but I think it's okay to include `@author` and `@version` into the class headers. Also, all your exceptions seem to print the same message. Perhaps you can consider `@Overriding` the `getMessage` method and replace it with a more relevant error message! Overall, good job. 😄

adamozh commented:
Thanks for the review and detailed feedback! @HangZelin

adamozh commented:
@CheongYeeMing Thanks for the detailed feedback! Didn't notice the Javadoc stuff 😮

adamozh committed:
Improve overall code quality

In some areas, Single Level of Abstraction Per method was violated,
making the code less readable. Calculations were abstracted out
into their own method, improving readability.

Other improvements to code quality:
* Renamed UI class to Ui to follow Java coding standards
* Fixed an else statement handling the last option
* Filled in missing JavaDoc parameters
* Standardised JavaDoc comments to end with a period
* Renamed difficult to read variables

Things to do:
* Make Parser non static
* Use streams
* Make 'bye' exit the application

adamozh committed:
Use Java Streams

In this case, it was already implemented. Using for loops in this
case would cause deep indentations due to loops and conditionals,
making the code less readable.

Streams were used to:
* Minimize indentation level
* Make code more expressive

adamozh committed:
Merge pull request #1 from moreTriangles/branch-A-Assertions

Add assert statements

adamozh committed:
Merge branch 'master' into branch-A-CodeQuality

adamozh committed:
Merge pull request #2 from moreTriangles/branch-A-CodeQuality

Improve overall code quality

adamozh committed:
Merge pull request #3 from moreTriangles/branch-A-Streams

Use Java Streams

adamozh committed:
Add friendlier syntax for some commands

Users had to type deadline, event and todo. Now, they can choose to
type d, e and t instead, saving them time. They can also access
list via l.

adamozh committed:
Merge branch 'branch-C-FriendlierSyntax'
Merge branch-C-FriendlierSyntax to master

adamozh committed:
Set Duke response to red during exceptions

adamozh committed:
Add Color.java to abstract away string hex codes

adamozh committed:
Add Javadoc for Color.java

adamozh committed:
Make Parser non static, Add some comments

adamozh committed:
Fix bug for Done and Delete when character entered

adamozh committed:
Fix checkstyle issues

adamozh committed:
Change User and Duke profile pictures

adamozh committed:
Delete old Duke and User profile pictures

adamozh committed:
Fix Javdoc and minor refactoring

adamozh committed:
Fix bug where empty list cannot be saved

adamozh committed:
Re-fix bug where empty file not saved

adamozh committed:
Change done symbol to a tick

adamozh committed:
Add Ui.png in docs

adamozh committed:
Update stage title to Duke

adamozh committed:
Update Ui.png

adamozh committed:
Update README.md for User Guide

adamozh committed:
Fix formatting for User Guide README.md

adamozh committed:
Fix more formatting for User Guide README.md

adamozh committed:
Fix more formatting for User Guide README.md

adamozh committed:
Fix Duke Image for user guide

adamozh committed:
Update response examples for done and delete

adamozh committed:
Add more Unit Tests

adamozh committed:
Update README.md

adamozh committed:
Update gradle file to inlcude dependencies in build

adamozh committed:
Update Command Summary

adamozh committed:
Add bye function to close Duke

adamozh committed:
Update README.md
